identity/src/main/java/com/shotcutter/identity/user/UserIdentityService.java

```diff
@@ -2,39 +2,117 @@
+                               UserRepository userRepository,
+                               ConverterService converterService,
+                               S3AsyncClient amazonS3) {
+        this.s3BucketName = s3BucketName;
         this.userRepository = userRepository;
         this.converterService = converterService;
+        this.amazonS3 = amazonS3;
     }
 
-    public Optional<UserEntity> findById(String id) {
+    public Mono<UserEntity> findById(String id) {
         return userRepository.findById(id);
     }
 
-    public Optional<UserEntity> findByEmail(String email) {
+    public Mono<UserEntity> findByEmail(String email) {
         return userRepository.findByEmail(email);
     }
 
-    public Optional<UserEntity> registerUser(User unregisteredUser) {
-        return converterService
-                .convertTo(unregisteredUser, UserEntity.class)
-                .map(userRepository::save);
+    public Mono<UserEntity> registerUser(User unregisteredUser) {
+        return Mono.justOrEmpty(converterService.convertTo(unregisteredUser, UserEntity.class))
+                .flatMap(userRepository::save);
     }
 
+    public Mono<UserEntity> updateAvatar(String userId, MultipartFile avatar) {
+        var path = getUserAvatarPath(userId, avatar);
+
+        return findById(userId)
+                .doOnNext(user -> {
+                    // if avatar for this user already exist at s3 - we need to remove it at first
+                    if (user.getAvatar() != null) {
+                        String avatarPath = null;
+                        try {
+                            avatarPath = new URL(user.getAvatar()).getPath();
+                        } catch (MalformedURLException e) {
+                            e.printStackTrace();
+                        }
+                        var relativeAvatarPath = avatarPath.substring(1);
+
+                        var deleteReq = DeleteObjectRequest
+                                .builder()
+                                .bucket(s3BucketName)
+                                .key(relativeAvatarPath)
+                                .build();
+
+                        amazonS3.deleteObject(deleteReq);
+                    }
+                })
+                .flatMap(user -> {
+                    var putObjReq = PutObjectRequest.builder()
+                            .bucket(s3BucketName)
+                            .acl(S3CannedAccessControlList.PUBLIC_READ.toString())
+                            .key(path)
+                            .build();
+
+                    CompletableFuture<PutObjectResponse> res = null;
+                    try {
+                        res = amazonS3.putObject(putObjReq, AsyncRequestBody.fromBytes(avatar.getBytes()));
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                    return Mono.fromFuture(res)
```

Comment:
placeholder for refactor: discussed in dm
we can zip userMono from 58 and mono for putting new avatar